### DISCUSSION TITLE
Will this change to FormController causes any Potential backward compatibility?

### DISCUSSION THREAD

**Discussion by joomdonation** [joomla/joomla-cms#32534] [General]
While working on this PR https://github.com/joomla/joomla-cms/pull/32520 to support Save As Copy feature for article in the frontend, I see that the FormController could not get ID of the article back after it is created from frontend (via FormModel class). The reason is:

- In the model class, we use the code below to store the ID of created record in state (name of model is used) https://github.com/joomla/joomla-cms/blob/4.0-dev/libraries/src/MVC/Model/AdminModel.php#L1415

```php
$this->setState($this->getName() . '.id', $table->$key);
```

- In the controller class, however, we use this code to get ID of the record back (basically, use name of controller class) https://github.com/joomla/joomla-cms/blob/4.0-dev/libraries/src/MVC/Controller/FormController.php#L752

```php
$recordId = $model->getState($this->context . '.id');
```
In most cases, it works well because we have model name the same with controller name. However, in case article is submitted from frontend, the model name is Form, not Article like the controller. The state name for storing record ID is **form.id**, not **article.id** as the controller expects

To solve that error, I changed the code in form Controller to

```php
$recordId = $model->getState($model->getName() . '.id');
```

To me, it is a bug fix but I'm unsure if it could cause any backward compatibility since we have the original code works like that for years. Ping @wilsonge , @laoneo and others for feedback.

**Comment by bembelimen** [comment]
```->getName()``` is the only correct call and it should rather fix bugs than raising new one as far as I see.

**Comment by laoneo** [comment]
Sounds like a valid change.

**Comment by wilsonge** [comment]
The context includes the component name. It stops things in the session from different components conflicting. So whilst I understand what change you need I don't think it will work :(

**Reply by joomdonation** [reply]
@wilsonge The $context variable contains component name and it remains unchanged. $this->context (which is used in that code) is name of the controller (not contains component name), see how it is detected here https://github.com/joomla/joomla-cms/blob/4.0-dev/libraries/src/MVC/Controller/FormController.php#L108-L127


Also, we use model name to store id of the record, so to get it back, we need to use model name, too, not controller name. So as other people said, this is a bug fix, not backward-incompatible change.

**Reply by wilsonge** [reply]
Well your going to loose all the active session data https://github.com/joomla/joomla-cms/blob/4.0-dev/libraries/src/MVC/Controller/FormController.php#L335 . So it's somewhat a b/c change. Why does this work in J3?

**Reply by joomdonation** [reply]
@wilsonge That session data is not related. Again:

- We store ID of the created record using this command https://github.com/joomla/joomla-cms/blob/4.0-dev/libraries/src/MVC/Model/AdminModel.php#L1415 . Please look at this line of code, we store ID of the record using model name

- Then in controller, we get it back using this https://github.com/joomla/joomla-cms/blob/4.0-dev/libraries/src/MVC/Controller/FormController.php#L752 . That's wrong. It works because in most case, $this->context is name of controller, and is the same with name of model.

Anyway, we store ID using model name, so we should get it back using model name, too.

In J3, we do not have Save As Copy feature from frontend, so we don't need to get ID of the created article back and that's the reason we do not see any error. 

Please look at the two lines of code I mentioned before, you will see that the change is valid and it's a bug fix (as @laoneo , @bembelimen and @Fedik agreed)
